pm: Refactor pm_notifier to work on a per-power-state basis

knthm opened the pull request:
The changes in this PR creates separate `pm_notifier` lists for each DT power state.
This allows more efficient traversal during power state transitions when many notifiers are needed, e.g. for driver-specific actions such as device reinitialization (see #59194 #60369).

The callback has also been refactored to no longer accept the (superficial) state that triggered it as an argument, but a transition indicator bit field and a void context pointer that can be used for a variety of purposes, such as passing a device driver instance.

#### Closes

-  #59194

knthm committed:
pm: Replace state transition bool with bit field

Add bit field values for specifying PM state transition direction, e.g.
state entry, exit, or both; use these in place of entering_state bool.

Move exising and add new pm_state_notify calls to the exact edges of
state transitions, to provide more flexibility and cover more use
cases.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
pm: Refactor pm_notifiers to be power-state specific

Replace individual entry and exit callbacks with a singular callback,
with a separate direction bit field to indicate whether the struct
applies to state entry/exit, or both.

Add a void pointer to provide optional context for the callback.

Replace the singular pm_notifiers slist with individual power state-
specific slists, generated from DT power-states definitions on SMP, and
cpu-power-states on single-CPU platforms.

Un/register functions now take a power (sub)state as a parameter, in
addition to the usual pm_notifier struct; used to place the notifier
into the correct list.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented:
@gmarull @ceolin please review at your earliest convenience :)

gmarull commented on include/zephyr/pm/pm.h:
```diff
@@ -53,16 +53,25 @@ extern "C" {
+	void (*callback)(void *arg1, void *arg2, void *arg3);
+
 	/**
-	 * Application defined function for doing any target specific operations
-	 * for power state exit.
+	 * Optional callback arguments
 	 */
-	void (*state_exit)(enum pm_state state);
+	void *arg1;
+	void *arg2;
+	void *arg3;
```
note here: I'd not repeat the ugly triple void * found in threads. I think a single `void *ctx` should be enough, or `const struct device *dev` if this API is expected to always be used in the context of devices.

knthm replied:
Good point, a single (const) void pointer would probably be sufficient.

I think the `pm_notifier` design was initially built to target application code, so I didn't want to directly add a device instance pointer, but leave it open for potential use by app code as well as drivers.

ceolin commented on include/zephyr/pm/pm.h:
```diff
@@ -53,18 +53,136 @@ extern "C" {
 	 * for power state entry.
 	 */
-	void (*state_entry)(enum pm_state state);
+	void (*callback)(void *arg1, void *arg2, void *arg3);
```
I would change it to something like:
```
void (*callback)(enum pm_state_direction direction, const void *ctx);
```

The reason to just having a single pointer for context is discussed bellow and there is no need to make this struct bigger for places where multiple args are not needed.

Giving the direction is something that I think makes sense, it allows re-using the same callback in case a user needs to do some work before/after a state. So, that means having something like `PM_STATE_ENTRY_AND_EXIT`

knthm replied:
Agreed on the `const void *ctx` part.

I can see the reasoning behind `PM_STATE_ENTRY_AND_EXIT`, but I think it muddies the purpose of the direction variable a bit, as we would never pass that third value to `pm_state_notify()`.

knthm replied:
Taking a second to think about it, I think adding `PM_STATE_ENTRY_AND_EXIT` makes sense.

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -156,6 +156,14 @@ struct pm_state_info {
+enum pm_state_direction {
+	PM_STATE_ENTRY,
+	PM_STATE_EXIT,
+};
```
Should't we have one for both ? If one needs to be notified before/after a state we can use only one callback, this can save space and timing iterating over the list

knthm replied:
In commit c4f8c1d6a63b68cdb91b77852ab5a82dcc218fd1 I moved/added additional `pm_state_notify()` calls to the edges of the respective state transitions: the exit of a low-power state is not exactly at the beginning of the active state, IMO:

https://github.com/zephyrproject-rtos/zephyr/blob/c4f8c1d6a63b68cdb91b77852ab5a82dcc218fd1/subsys/pm/pm.c#L152-L158

For my use case within a driver for reinit, I'd want to the callback to trigger before the `pm_state_exit_post_ops()` calls, while an application would probably want it to trigger after it. Hence the explicit distinction between low-power exit and active-state entry seen here.

ceolin replied:
got that, saying that if you want to do something before/after `SOME_IDLE_STATE` you could use something like `PM_STATE_ENTRY_AND_EXIT`. For you case you use `PM_STATE_EXIT`, nothing changes. 

Note, that basically any state entry is another state leaving so what I am saying could be done with two notifiers, one when leaving active and other when leaving idle but having both directions option allow us to use only one calback.

knthm replied:
Ohh I see, it touches on your similar comment regarding the callback parameters. Sorry, misunderstood you. :)

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -339,6 +347,18 @@ struct pm_state_info {
+ */
+uint8_t pm_state_cpu_get_index(uint8_t cpu, enum pm_state state,
+			       uint8_t substate_id);
+
```
I don't think this needs to be part of the API tbh, if really needed this could probably be internal to where it is needed.

knthm replied:
I thought it might be interesting to have. The longer I looked at usage of `z_cpus_pm_state[id].state`, the more I thought it might be a better solution to have a pointer to a power state array member inside it instead of just an unlinked `pm_state_info` instance.

knthm replied:
Removed, as no longer applicable.

ceolin replied:
This function should return a signed int since it returns a negative value in case of error, otherwise we will have -1 wrongly promoted to a positive value.

Also, my point is, that is a very very specific function, lets not include it in a public header, just remove this declaration and declare it in  `pm.c`

``extern int8_t pm_state_get_index(enum pm_state state, uint8_t substate_id);``

ceolin commented on include/zephyr/pm/pm.h:
```diff
@@ -53,18 +53,136 @@ extern "C" {
+/**
+ * @brief Generate individual, per power-state pm_notifiers lists for the
+ *	  passed CPU id.
+ *
```
It shouldn't be per CPU imho, if you have a list per CPU you will probably need to have N notifiers copy (one per cpu). Unless you have specific needs per cpu what you really want to know is when a state is used and in this case  you would need to have a copy per cpu.

knthm replied:
I'll do some more research into Zephyr's SMP implementation, but I think I understand your point.

What do you think of a CPU-independent array of lists per `power-states` node? That would also significantly simplify the macros added here, and should avoid SMP issues.

ceolin replied:
Yep, that is my point. Just keep lists per states (or one single list if this is not really a bottleneck)

knthm replied:
Done.

ceolin commented on subsys/pm/pm.c:
```diff
@@ -255,23 +305,62 @@ bool pm_system_suspend(int32_t ticks)
+	k_spinlock_key_t pm_notifier_key;
+	sys_slist_t *notifier_list;
+	uint8_t id = CURRENT_CPU;
+
```
I've added a comment about registering per cpu, but here is another possible problem. In SMP you don't know which cpu you are running (Unless you have been pinned). So, the caller would need to check if it is the right CPU, sets a spin lock and call it ...  It is even worse if you are not running from the CPU you wants to get notification

knthm replied:
TBF I don't have a very deep understanding of how SMP works in Zephyr, since it seems a bit different from Linux with the MCU configurations of e.g. M4 + M7 cores etc.

From a quick glance I can see that Zephyr on x86 with SMP will in fact run across multiple CPUs. Do you know if this is the same on systems consisting of an M4 and M0+ core, for instance? My design tried to cover these cases, as the different cores may have diverging `cpu-power-states`, Zephyr may be pinned to the main M4 core, etc. (These are probably very flawed assumptions).

ceolin replied:
if the core does not have a state, it will not be triggered during suspension.

knthm replied:
Understood, thanks for the clarification. Functionality is now CPU-independent.

ceolin commented:
@knthm Thanks for putting it together. Some high level comment

- Is running the notification list a performance bottleneck ? If it is just a guess, I would have an initial version simplified that has a single list (as before) instead of per cpu (that contains some issues) and state.
  -  This is an internal detail that does not change any interface, so if performance become a problem we can have it more elaborated. My concern is that we are having all these macros and more complexity for a few items in a list.

knthm commented:
Thanks for the review!

> I would have an initial version simplified that has a single list (as before)

That's a fair opinion. Maybe we could compromise on a CPU-independent per-`power-states` array of lists?
I'll push the commits when I find a minute these next few hours.

knthm commented:
- Replaced three superfluous void pointers with singular `const void *ctx`
- Added `PM_STATE_ENTRY_AND_EXIT` direction
- Removed now non-applicable `pm_state_cpu_get_index` API
- Simplified `pm_notifier` slist generation: slists are now generated per general power-state, independent of any CPU and their individual `cpu-power-states`

knthm commented on subsys/pm/pm.c:
```diff
@@ -27,7 +27,24 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
+static sys_slist_t pm_notifiers_active = SYS_SLIST_STATIC_INIT(&pm_notifiers_active);
+
+/* Generate a notifier list for each DT power state */
+DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states), PM_NOTIFIERS_CREATE_STATES, (;));
```
Should I wrap this in a `DT_NODE_EXISTS` check, in case the `cpus` node doesn't have a `power-states` child?

knthm replied:
Tested: DT errors will occur if `power-states` isn't defined. Adding `COND_CODE_1` wrapper

gmarull commented on include/zephyr/pm/state.h:
```diff
@@ -156,6 +156,15 @@ struct pm_state_info {
+ * @enum pm_state_direction Entry/exit indicator for state transitions
+ */
+enum pm_state_direction {
+	PM_STATE_ENTRY,
+	PM_STATE_EXIT,
+	PM_STATE_ENTRY_AND_EXIT,
```
I think this should just be a bit field, that you can OR if you need both.

knthm replied:
That'd indeed be more elegant. I'll try that out, thanks.

ceolin replied:
yep, that is better indeed.

knthm replied:
Done.

gmarull commented on include/zephyr/pm/pm.h:
```diff
@@ -53,16 +53,23 @@ extern "C" {
 	 * for power state entry.
 	 */
-	void (*state_entry)(enum pm_state state);
+	void (*callback)(enum pm_state_direction direction, const void *ctx);
```
nit: document callback arguments

also, consting context may be unnecessarily restrictive.

knthm replied:
Good point, will add argument documentation.

The context is const for the use case of a driver of passing a device struct.

ceolin replied:
yep ... you keep the argument const in the internal struct and remove the const when calling the callback,

knthm replied:
I understand that, but doesn't the potential for undefined behavior make this problematic? Though I suppose in a driver we'd immediately cast ctx into a `const struct device *`.

gmarull commented on include/zephyr/pm/pm.h:
```diff
@@ -72,6 +72,31 @@ struct pm_notifier {
+ *
+ * @param node_id Power states child node identifer
+ */
+#define Z_PM_NOTIFIERS_NAME(node_id) UTIL_CAT(pm_notifiers_, DT_NODE_CHILD_IDX(node_id))
+
+/**
+ * @brief Generate a pm_notifiers list for the passed power-states child node
+ *	  identifier.
+ *
+ * @param node_id CPU node identifier
+ */
+#define PM_NOTIFIERS_CREATE_STATES(node_id)				\
```
all DT specific APIs should contain `_DT_` to make it explicit that it has to be used with DT

also, why is this one exposed publicly?

knthm replied:
Added `_DT_` and `Z_` prefix.

gmarull commented on subsys/pm/pm.c:
```diff
@@ -27,7 +27,24 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
+
+/* Array of all DT power-states, active state as first element */
+static const struct pm_state_info pmstates[] = {
+	{ .state = PM_STATE_ACTIVE },
```
question: why is active needed here?

knthm replied:
Active states are included in entry/exit notifications.

Here specifically, it mirrors the array contents and order of `pm_notifiers_array[]`.

Prepending the active state to both lets us use the index to refer to the same state between either, this is used to compare against the current system state and execute the correct notifier list, as well as during notifier un/registration.

gmarull commented on include/zephyr/pm/pm.h:
```diff
@@ -97,6 +97,33 @@ struct pm_notifier {
  */
 #define PM_NOTIFIERS_STATES(node_id) &Z_PM_NOTIFIERS_NAME(node_id)
 
+#ifdef CONFIG_PM
+/**
+ * @brief Application/driver helper for defining an instance-specific
+ *	  pm_notifier struct.
+ *
+ * @param index Application/driver instance index
+ * @param state_direction pm_state_direction: state entry, exit, or both
+ * @param callback_fn callback function
+ */
+#define PM_NOTIFIER_INST_DEFINE(index, state_direction, callback_fn)	\
+static struct pm_notifier pm_notifier_##index = {			\
+	.direction = state_direction,					\
+	.callback = callback_fn,					\
+	.ctx = DEVICE_DT_INST_GET(index),				\
+}
+
+/**
+ * @brief Retrieve the name of a pm_notifier struct instance.
+ *
+ * @param index Application/driver instance index
+ */
+#define PM_NOTIFIER_INST(index) pm_notifier_##index
```
ditto for naming (`_DT_`).

What about drivers not using instance indices? `DT_INST` macros are not always used by drivers. I'd suggest having a non-inst version (ie taking node_id) and a inst version, expanding to the right node_id from an instance. You can use the unique DT_ORD number internally for naming purposes.

knthm replied:
Note that these helper macros aren't directly devicetree dependent, as they also need to work from application code. Index in this case can be anything, A device instance index, ORD number, or any other string/number.

Should I replace the `_INST_` part of the name to remove the confusion?

knthm replied:
I see I forgot to push my change that removes the use of `DEVICE_DT_INST_GET` from said macro. Above comment will apply after that.

gmarull replied:
> Note that these helper macros aren't directly devicetree dependent, as they also need to work from application code. Index in this case can be anything, A device instance index, ORD number, or any other string/number.
> 
> Should I replace the `_INST_` part of the name to remove the confusion?

Then we must be careful, because we can have potential collisions. I'd have a generic API, and a device specific variant built on top, that uses ordinals as its internal naming scheme. The generic should document avoidance of numeric only names.

ceolin replied:
This should be generic, we shouldn't assume that the context is always a device, make it a parameter and probably replace index with some other identifier (like _name).

knthm replied:
Added generic function and DT node/instance variations that make use of it.

Let me know if you find the implementation acceptable.

gmarull reviewed (dismissed):
thanks for this proposal, left some comments

knthm committed:
tests: Update tests/samples affected by pm_notifier API change

Update pm tests and sample that make use of pm_notifier functionality,
or test system power state transitions.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented on tests/subsys/pm/power_mgmt/src/main.c:
```diff
@@ -181,11 +181,13 @@ void pm_state_set(enum pm_state state, uint8_t substate_id)
+	/* at this point, notify_pm_state() implemented in
 	 * this file has been called and set_pm should have been set
 	 */
+	/*
 	zassert_true(set_pm == true,
 		     "Notification to enter suspend was not sent to the App");
+	*/
```
For some reason, `pm_set_state` is triggered during test_device_order, before any notifiers are registered?
Thus said test fails due to the notifier not having set this bool.

ceolin commented on include/zephyr/pm/pm.h:
```diff
@@ -53,18 +53,119 @@ extern "C" {
+ * @param inst Instance number
+ */
+#define PM_NOTIFIER_DT_INST(inst) PM_NOTIFIER(DT_INST_DEP_ORD(inst))
+
```
I particularly don't see a reason to have this specialization here, if one needs using it in devices, a notifier with the device as context can easily be created with the device. Also, these macros are confusing IMHO since they don't tell anything about having a device as context.

knthm replied:
I suppose that applies to all the `_DT` macros added here. IMHO they help driver/app code look cleaner, though naming could be improved.

Would something like `PM_NOTIFIER_DT_DEVICE_INST` provide enough information that the device is passed/present in context?

knthm replied:
Adapted macro naming, please have a look.

ceolin replied:
What I was saying is to have only `PM_NOTIFIER_DEFINE` and let what is done in these macros to the driver to do. Further if that become re-current we introduce these helpers, I would avoid adding more APIs than needed for now.

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -156,6 +156,16 @@ struct pm_state_info {
+/**
+ * @brief PM state transition direction: state entry bit field value
+ */
+#define PM_STATE_ENTRY        (1 << 0)
```
Use the macro `BIT` for this.

ceolin commented on subsys/pm/pm.c:
```diff
@@ -107,26 +136,40 @@ static void pm_resume_devices(void)
 	struct pm_notifier *notifier;
 	k_spinlock_key_t pm_notifier_key;
-	void (*callback)(enum pm_state state);
+	uint8_t id = CURRENT_CPU;
```
leftover ?

knthm replied:
The `id` var? I use that on the line below to retrieve the current system state. I thought it was a bit more readable that way (at the cost of a byte of stack mem).

ceolin commented on subsys/pm/pm.c:
```diff
@@ -27,7 +27,36 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
+		(DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states),
+						  PM_STATE_INFO_DT_INIT, (,))),
+		())
+};
```
You don't need it, that is pretty much what you get in `pm_state_cpu_get_all()`.

ceolin replied:
good point about per cpu. I don't see any problem adding this API to return all defined states.
We should just try to avoid duplicate this data, specially in single core cases, that the list is exactly the same of cpu0.

knthm replied:
Sounds good. To avoid duplication, I could wrap the logic in `#if CONFIG_MP_NUM_CPUS != 1` or similar.

ceolin commented on subsys/pm/pm.c:
```diff
@@ -27,7 +27,36 @@ LOG_MODULE_REGISTER(pm, CONFIG_PM_LOG_LEVEL);
+	&pm_notifiers_active,
+	COND_CODE_1(
+		DT_NODE_EXISTS(DT_PATH(cpus, power_states)),
+		(DT_FOREACH_CHILD_STATUS_OKAY_SEP(DT_PATH(cpus, power_states),
```
Maybe its cleaner to have something like:
```
#if DT_NODE_EXISTS(DT_PATH(cpus, power_states))
...
#else
static sys_slist_t *pm_notifiers_array[] = {&pm_notifiers_active}; 
#endif
```

ceolin commented on subsys/pm/pm.c:
```diff
@@ -107,26 +136,39 @@ static void pm_resume_devices(void)
 #endif	/* CONFIG_PM_DEVICE */
 
+/* Retrieve the element index of a power state. */
+uint8_t pm_state_get_index(enum pm_state state, uint8_t substate_id)
```
`static uint8_t pm_state_get_index(enum pm_state state, uint8_t substate_id)`

knthm replied:
Good call.

On a side note: any thoughts on this and the `static const struct pm_state_info pmstates[]` above living in `state.{c,h}` and becoming part of the API? I think there's likely some use cases outside of the notifier stuff for it.

ceolin replied:
yep (look my comment about data duplication). Another thing, this function is a good candidate to be inlined.

ceolin reviewed (changes requested):
129b50dbf6d927adf4f657b8ad4315b40b5f33d9 has to be squashed with f190b2e69b4ec86e64c4a70688396dfc05cb46ea to not break the tree.

knthm committed:
pm: state: Add power-states array index helpers

Add pmstates array generated from global power-states DT node children,
vs. current per-CPU only pmstates array (SMP platforms only).

Add helper functions for retrieving the index of a global/per-CPU
power-state array element.
These functions are intended to faciliate easier processing of similar
PM data structures generated from the same DT power-states nodes.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented on subsys/pm/state.c:
```diff
@@ -40,12 +40,20 @@ BUILD_ASSERT(DT_NODE_EXISTS(DT_PATH(cpus)),
 
+#if CONFIG_MP_NUM_CPUS > 1 && DT_NODE_EXISTS(DT_PATH(cpus, power_states))
+/** General power-states list on SMP platforms, as individual per-CPU states may vary */
+static const struct pm_state_info pmstates_all[] = {
```
This guard condition turns out to be somewhat problematic:
In my use case #59200, a state can be defined in `power-states` and `reinit-power-states`, but be missing from `cpu-power-states`. This'll cause a crash when we register a notifier for that missing cpu power state, as we'd try to reference a notifier which doesn't exist.

In summary, we kind of need both lists and shouldn't exclude pmstates_all on single core platforms :/

We

erwango replied:
To be sure IIUC, `stop2` is `reinit-power-states` on usart1, as this is a configuration and not in `cpu-power-states` because this is your application configuration. Am I correct ?

knthm replied:
Yes, this is just a corner case, but I think we should cover it as the resulting errors aren't entirely obvious. What do others think though?

knthm replied:
Any thoughts @ceolin & @gmarull?

knthm replied:
Turns out [I'm an idiot](https://github.com/zephyrproject-rtos/zephyr/pull/60998/files#diff-d4e51fe3d6c0c3c82cef6a93687ea6246bc506c89f069f071138be1d497459d6R143-R145).

knthm commented:
- Rebased on main
- Fix underflow condition on `state_idx` variable

ceolin commented on include/zephyr/pm/pm.h:
```diff
@@ -87,8 +156,13 @@ bool pm_state_force(uint8_t cpu, const struct pm_state_info *info);
  */
-void pm_notifier_register(struct pm_notifier *notifier);
+int pm_notifier_register(struct pm_notifier *notifier, enum pm_state state, uint8_t substate_id);
 
```
I am thinking about it, wouldn't be better having notifier only per state (and not state + substate). I wondering that including substate will make really hard re-using the driver across different SoCs. The callback though will include state / substate  .... or maybe one solution is to use the information about which states should do device pm (from  #60939) or something similar.  Obviously, that can be done later, that is why we can have a simple version of this pr in first.

knthm replied:
It is intended that the driver obtains the `power-state` directly from DTS via macros, so the driver will always have the correct state/substate combination. The state/substates combinations aren't intended to be hardecoded.

See my recommended implementation in the driver [here](https://github.com/zephyrproject-rtos/zephyr/pull/59200/commits/3931c014abfe45d0e93f6ab3999875e4185268d5#diff-54bdff4ff3a16e194201b66e24cada279fdfdd40dfd66b1df3df3a99d0f65cd0R2199-R2200).

erwango replied:
Agree with @knthm. @ceolin would you approve the current state then ?

knthm commented:
ping @gmarull

erwango commented:
> @[erwango](https://github.com/erwango) erwango requested a review from [gmarull](https://github.com/gmarull) [1 minute ago](https://github.com/zephyrproject-rtos/zephyr/pull/60998#event-10461738467)

Clicked too fast, my bad.

knthm commented on subsys/pm/pm.c:
```diff
@@ -149,10 +175,13 @@ void pm_system_resume(void)
+		/* Exiting low-power state */
+		pm_state_notify(PM_STATE_EXIT);
 		pm_state_exit_post_ops(z_cpus_pm_state[id].state, z_cpus_pm_state[id].substate_id);
-		pm_state_notify(false);
```
Note to self:

On STM32 we need to reinitialize peripherals after we've reinitialized clocks, but before we re-enable interrupts.
Both of these occur inside of `pm_state_exit_post_ops`, thus we should probably rethink the approach here.

ceolin replied:
Isn't the case of re-initialize clocks in `pm_state_set` just before leaving this function. e.g:
```

/* Invoke Low Power/System Off specific Tasks */
void pm_state_set(enum pm_state state, uint8_t substate_id)
{
	if (state != PM_STATE_SUSPEND_TO_IDLE) {
		LOG_DBG("Unsupported power state %u", state);
		return;
	}

	/* ensure the proper wake-up system clock */
	LL_RCC_SetClkAfterWakeFromStop(RCC_STOP_WAKEUPCLOCK_SELECTED);

	switch (substate_id) {
	case 1: /* this corresponds to the STOP0 mode: */
		/* enter STOP0 mode */
		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
		break;
	case 2: /* this corresponds to the STOP1 mode: */
		/* enter STOP1 mode */
		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
		break;
	case 3: /* this corresponds to the STOP2 mode: */
#ifdef PWR_CR1_RRSTP
		LL_PWR_DisableSRAM3Retention();
#endif /* PWR_CR1_RRSTP */
		/* enter STOP2 mode */
		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP2);
		break;
	default:
		LOG_DBG("Unsupported power state substate-id %u",
			substate_id);
		break;
	}

	LL_LPM_EnableDeepSleep();
	/* enter SLEEP mode : WFE or WFI */
	k_cpu_idle();

       /* We have just waken up from idle, re-initialize clock */
	stm32_clock_control_init(NULL);
}
```

Then you have notification callback after the clock being initialized and before restoring interruptions.

knthm replied:
Moving clock reinit to after idle in `pm_state_set` on STM32 won't have the intended affect, as the `pm_system_resume()` logic (called via idle's `z_pm_save_idle_exit()`) is run in the ISR of the wakeup source, and not in the idle thread.

I was thinking of refactoring the IRQ unlocking: it's pretty unnecessary to do this in the `pm_state_exit_post_ops()`, as *every* platform needs to reenable IRQs after wakeup to reverse the `irq_lock()` done by idle.
On several platforms the post_ops are used exclusively for reenabling IRQs, so maybe extending the API to could deal with it on the architecture-level instead of for each SoC-series could be a good way to remove this redundancy. It'd also allow us to wedge the notifier callback in between the post_ops and irq_unlock.

knthm replied:
For now I've moved the notifier to after post_ops. The IRQ changes are larger as there are discrepancies in some post_ops logic that block it from being implemented cleanly. I'll create a draft PR for it when ready.

ceolin commented on include/zephyr/pm/state.h:
```diff
@@ -339,6 +361,27 @@ struct pm_state_info {
+ *
+ * @return Index of the PM state, or -1.
+ */
+uint8_t pm_state_cpu_get_index(uint8_t cpu, enum pm_state state, uint8_t substate_id);
```
this function is not needed, the only one used is `pm_state_get_index`

knthm replied:
Nice catch. I think we should probably drop `pm_state_get_index` and use the CPU-specific one instead, just to match [the call here](https://github.com/zephyrproject-rtos/zephyr/pull/60998/files#diff-d4e51fe3d6c0c3c82cef6a93687ea6246bc506c89f069f071138be1d497459d6R143-R144) when we pass the CPU-specific sub/state for which we want to retrieve the index.
On single-core platforms the `#if`-guarded variant would just drop the CPU id. Otherwise, its current state this could lead to discrepancies and notifiers not triggering if cores have diverging `cpu-power-states` definitions.

ceolin commented on samples/boards/mec15xxevb_assy6853/power_management/src/power_mgmt.c:
```diff
@@ -256,21 +255,24 @@ static void resume_all_tasks(void)
-	.state_exit = notify_pm_state_exit,
-};
+PM_NOTIFIER_DEFINE(test_light_sleep, (PM_STATE_ENTRY|PM_STATE_EXIT), notify_pm_state, NULL);
+PM_NOTIFIER_DEFINE(test_deep_sleep, (PM_STATE_ENTRY|PM_STATE_EXIT), notify_pm_state, NULL);
```
these variables names could have something that say that they are notifiers. We have
`test_light_sleep` and `light_sleep` variables in this file

ceolin commented on subsys/pm/state.c:
```diff
@@ -66,3 +74,34 @@ uint8_t pm_state_cpu_get_all(uint8_t cpu, const struct pm_state_info **states)
+}
+
+#if CONFIG_MP_NUM_CPUS > 1
+inline uint8_t pm_state_get_index(enum pm_state state, uint8_t substate_id)
```
see comments about returning type

knthm replied:
Addressed in response above.

knthm commented:
- Rebased on main
- Addressed comments left by @ceolin

knthm commented:
- Rebased on main
- Addressed comments left by @ceolin

knthm commented:
- Rebased on main
- Addressed comments left by @ceolin

tmleman commented on subsys/pm/pm.c:
```diff
@@ -176,6 +179,9 @@ bool pm_system_suspend(int32_t ticks)
 
 	SYS_PORT_TRACING_FUNC_ENTER(pm, system_suspend, ticks);
 
+	/* Exiting active state */
+	pm_state_notify(PM_STATE_EXIT);
+
```
At this point you don't know yet that you will go to any of the lower power states. Notification seems to be to early.

tmleman reviewed (changes requested):
Currently, I see no reason to add notifications about leaving the ACTIVE state. Each transition to a low power state is equivalent to this. Why notifications about entering a lower state is not enough?

shilga commented:
Hi,
what is blocking this PR at the moment? This and other related PRs are essential to get anything meaningful done on stm32wl series.
